docs/src/rfcs/002-fearless-concurrency.md

```diff
@@ -0,0 +1,220 @@
+# Fearless Concurrency (V2.0)
+
+
+## Summary
+
+GPU programs are executed across thousands or millions of independently executing threads, grouped into a hierarchy of [scopes](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/chap9.html#shaders-scope).
+
+Unlike threading on a CPU, threads on a GPU cannot easily talk to each other - there are limitations on how a group of threads can be synchronized, and how. Values written by one thread are also not automatically visible to other threads with acquire and release operations like on a CPU. With threads grouped into scopes, applications must be explicit about which scopes threads are synchronized within, and to what scope writes become visible. Both of these requirements must be carefully navigated in order to avoid data races.
+
+In GLSL and HLSL (see [Prior art](#prior-art)), applications are explicit about synchronization and visibility, but lack the safety that Rust typically affords. This proposal introduces functions which are also explicit, but integrate into rust's borrow checking to provide safety against data races. This proposal does not deal with storage images, atomic operations, subgroup operations, or any more advanced functionality.
+
+This proposal currently requires the use of the `VulkanMemoryModel` capability, and describes everything in terms of that - however it could be mapped to SPIR-V without this capability in future if that's desirable.
+
+For now, this proposal only deals with barriers and the Workgroup scope as it affects buffer storage and workgroup storage; other scopes and atomics etc. are left to future proposals.
+
+## Explanation
+
+This is a baseline proposal which is expected to be built on top of over time, to incorporate new ideas and more flexibility.
+
+### Resource Types
+
+Storage buffer data is wrapped with the following type:
+
+```rust
+struct MutDeviceStorage<T> {
+    data: StorageBuffer<T>,
+};
+```
+Workgroup storage is wrapped with a single type:
+
+```rust
+struct MutWorkgroupStorage<T> {
+    data: Workgroup<T>,
+};
+```
```

Comment:
Nit: In rust the convention is put `Mut` at the end of the type name. So these would be `DeviceStorageMut` and `WorkgroupStorageMut` respectively.

Some examples:
- https://doc.rust-lang.org/stable/std/collections/btree_map/struct.IterMut.html
- https://doc.rust-lang.org/stable/std/ops/trait.DerefMut.html
- https://doc.rust-lang.org/stable/std/borrow/trait.BorrowMut.html

Reply:
Ah yes, I will update. Thanks!